services/supervisor/src/transactions.rs

```diff
@@ -333,49 +350,90 @@ impl SupervisorInterface for Supervisor {
+                    }
+                    consensus_propose_added = true;
+                    config
+                        .validate()
+                        .map_err(|e| (Error::MalformedConfigPropose, e))?;
+                }
 
-        // Verifies that the instance name does not exist.
-        if context
-            .data()
-            .for_dispatcher()
-            .get_instance(service.name.as_str())
-            .is_some()
-        {
-            return Err(Error::InstanceExists.into());
+                ConfigChange::Service(config) => {
+                    if !service_ids.check_unique(config.instance_id) {
+                        log::trace!("Discarded multiple service change proposals in one request.");
+                        return Err(Error::MalformedConfigPropose.into());
+                    }
+
+                    context
+                        .interface::<ConfigureCall<'_>>(config.instance_id)?
+                        .verify_config(config.params.clone())
+                        .map_err(|e| (Error::MalformedConfigPropose, e))?;
+                }
+
+                ConfigChange::StartService(start_service) => {
+                    if !services_to_start.check_unique(start_service.name.clone()) {
+                        log::trace!(
+                            "Discarded multiple instances with the same name in one request."
+                        );
+                        return Err(Error::MalformedConfigPropose.into());
+                    }
+
+                    start_service.validate(&context)?;
+                }
+            }
         }
+        Ok(())
+    }
+}
 
-        let confirmations = Schema::new(context.service_data())
-            .pending_instances
-            .confirm(&service, author);
-        let validator_count = core_schema.consensus_config().validator_keys.len();
-        if confirmations == validator_count {
-            log::trace!(
-                "Request add service with name {:?} from artifact {:?}",
-                service.name,
-                service.artifact
-            );
-            // We have enough confirmations to add a new service instance;
-            // if this action fails this transaction will be canceled.
-            context.start_adding_service(service.artifact, service.name, service.config)?;
+/// A helper structure designed to check whether a newly
+/// introduced element is unique among others.
+#[derive(Debug, Default)]
+struct UniqueSet<T: std::hash::Hash + Eq + Default> {
+    inner: HashSet<T>,
+}
+
+impl<T: std::hash::Hash + Eq + Default> UniqueSet<T> {
+    /// Creates a new `UniqueSet` object.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Checks whether element is unique, and if so
+    /// adds it to the collection.
+    pub fn check_unique(&mut self, element: T) -> bool {
+        if self.inner.contains(&element) {
+            return false;
         }
 
-        Ok(())
+        self.inner.insert(element);
```

Comment:
Nit: [`HashSet::insert()`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert) returns whether the inserted element is new, so the `contains()` call above can be eliminated.